File: CodeEval/hard/014-String-Permutations/solution.py
#!/usr/bin/env python

from sys import argv

def permutations(word):
    if len(word) == 1:
        return [word]
    first = word[0]
    perms = permutations(word[1:])
    result = []
    for perm in perms:
        for i in range(len(perm)+1):
            result.append(perm[:i] + first + perm[i:])
    return result

for line in open(argv[1], 'r'):
    line = line.strip()
    if line == "":
        continue
    print(",".join(sorted(permutations(line))))

File: CodeEval/hard/014-String-Permutations/solution.rb
#!/usr/bin/env ruby

def permutations(word)
  return [word] if word.length == 1
  perms = permutations(word[1..-1])
  result = []
  perms.each do |perm|
    (perm.length+1).times do |i|
      result.push("#{perm[0...i]}#{word[0]}#{perm[i..-1]}")
    end
  end
  return result
end

File.open(ARGV[0]).each_line do |line|
  line.strip!
  next if line.empty?
  puts permutations(line).sort.join(",")
end
